grafana/grafana
# Prometheus: Run exemplars explore queries through backend

**What this PR does / why we need it**:
Part of Prometheus migration to backend data source #37784

This PR focuses on making Explore + Exemplars queries run trough backend. It means that now all Explore queries are run trough backend. I have left comments in this PR to explain specific parts right next to the code,  but to summarize, this PR:
**Backend:**
1.  Moves sampling and processing of exemplars to Data Frames to backend. The sampling is identical to frontend sampling and if you run 1 query trough backend and 1 trough frontend, the results are same.
1. Removes passing of query type from backend to frontend and rather passes result type (vector/scalar/exemplar/matrix).
1. Adds tests for backend processing of exemplars.

**Frontend:**
1. Changes and unifies the way we are knowing if exemplars query option should be enabled or disabled. ((previously it wasn't showing disabled exemplars in Explore instant queries.)
3. Refactors `PromExploreExtraField` so we have all methods in there rather than passing it from `PromExploreQueryEditor`(slightly out of scope, could be moved to separate PR. if requested)
4. We are only sending exemplar query if the query has histogram metrics. Previously, we were always sending it which could be redundant as exemplars are available only for histogram metrics.  


**To test this:**
Running data source with exemplars:
1. Clone `https://github.com/grafana/tns` 
2. `cd tns/production/docker-compose` 
3. Run `docker plugin install grafana/loki-docker-driver:latest --alias loki --grant-all-permissions` and then `docker-compose up -d`
4. Prometheus data source on `localhost:9090` has `exemplars` and you can use them for testing.
5. To see exemplars, use this query: `histogram_quantile(.99, sum(rate(tns_request_duration_seconds_bucket{}[1m])) by (le))`

Running data source that supports exemplars, but doesn't have it - `http://demo.robustperception.io:9090/`
Running data source that doesn't support exemplars - update version in `grafana/devenv/docker/blocks/prometheus/Dockerfile` to smoething older e.g. `FROM prom/prometheus:v2.11.2`

When I was testing if the exemplars are same, in code, I added condition to `public/app/plugins/datasource/prometheus/datasource.ts` line 320 `shouldRunBackendQuery` that when target has both - range & instant - run exemplars trough frontend, if it only has range, run it trough backend and this way I was able to compare it in split pane.  
E.g. here:
![image](https://user-images.githubusercontent.com/30407135/136397359-07d61fbc-d609-4493-8eed-ba66e6a5f634.png)
